### ISSUE TITLE
relation extraction task

### ISSUE THREAD

**Issue by JehadAldahdooh** [dmis-lab/biobert#133]
I have some questions on relation extraction task.
As an example in train.tsv:
@GENE$ methylation occurs frequently in human colonic @DISEASE$ and cancers and is closely associated with K-ras mutations.        0

### What does label mean here 0?

--------------------------
In test.tsv:
index   sentence          label
0          Coupled with previous findings, our data suggest that heritable @GENE$ and IL10 variations may contribute to the acquisition or progression of @DISEASE$ and that the effects of other targeted loci in the cytokine and chemokine system cannot be established unequivocally in the study populations.      1

### Why should we need to provide a label in test.tsv? It should be hidden and Biobert task is to predict it?

-------------------------
### What is the task name for multi-class data should be?

 I know you use CHEMPROT data but Could you please provide me how the input/test format should be? 

look forward to your reply.

Thanks a lot

**Comment by AndreasSaka** [comment]
@JehadAldahdooh I am a fellow researcher. I tried sending you some hints to your email. Can you send me your email again bcs for some reason I could not send you the answers? Kind regards.

**Comment by JehadAldahdooh** [comment]
@AndreasSaka  Thanks a lot for your reply. I have received your email and would happy to continue discussing with you.

Thanks a lot 👍

**Comment by vierzh** [comment]
Hi, I have the same problem. Could you tell me how you solved the problem? 
Thanks!

**Comment by JehadAldahdooh** [comment]
Hi what kind of problem do you have?

Please send me an email and I would be happy to help you.

Thanks a lot

**Comment by vierzh** [comment]
Thanks for your reply. I have the same confusion on the question you asked, such as the meaning of label 1 and label 0, the reason of providing a label in test.tsv, and the input/test format of multi-class data. 
My email：1831559@tongji.edu.cn

Really appreciate your help!

**Comment by LedaguenelArthur** [comment]
Hi, I would be interested as well by the answers to the previously asked questions (i.e. signification of the label and presence of the label in the test file), as I struggle to understand how to fine-tune BioBERT on Relation Extraction.
In addition, I noticed that the last commit on run_re is named "Update : RE (chemprot only)" : does this mean that the script can now only be used on chemprot dataset ? or simply that only chemprot preprocessing class was impacted by the commit ?

If my email is needed : arthur.ledaguenel@student.ecp.fr

Thx a lot in advance !

**Comment by wangxinyi-gsafety** [comment]
@JehadAldahdooh, 
I have the same confusion on the question you asked, such as the meaning of label 1 and label 0, the reason of providing a label in test.tsv, and the input/test format of multi-class data such as CHEMPROT.

I would like to know how to run the run_re.py script for chemprot dataset. CHEMPROT dataset is a multi-class classification dataset, I simply run the run_re.py script with the same format like "gad" task, but error happened with no train.tsv file found.

my email is: wangxinyi@gsafety.com

looking forward to your reply and thanks.

**Comment by JehadAldahdooh** [comment]
Hi All,

This is a follow-up emails regarding my question and some explanations which I hope you find it very useful:

**Andreas**: 

I read your comment for the relation extraction part! I found online in scibert repo the chemprot dataset a while ago. In general label 0 mean there is no relation between the predefined tags and label 1 means there is a relation between the candidate predefined tags

The BioBERT performs the relation classification task and not relation extraction per se. The result, in a sentence level, assumes that a sentence co-occurrence of two entities signals a positive relation. To perform that on your raw sentences you need to recreate the input, meaning that you need a test.tsv file with index, sentence, and a label between values [0,1]. You will also need train.tsv, dev.tsv. The dev.tsv should be an empty file.

**### Me:**
The BioBERT performs the relation classification task and not relation extraction per se.

So it classifies the sentence either it has a relation or not, but what I see from the Biobert results. It looks like:
0.10590795     0.14545755     0.15841042     0.16602148     0.11944838     0.1241328       0.18062152
0.10492526     0.15121962     0.15111274     0.17243096     0.13778538     0.119106516   0.16341946

For the following labels: 

  def get_labels(self):
    """See base class."""
    return ["cpr:3", "cpr:4", "cpr:5", "cpr:6", "cpr:9","cpr:23", "false"]

so, if I want to clarify the results compared to the labels. I can understand that the probabilities in the first result line would mean the class false and in the second line would mean cpr:6 

In this case, it is pretty clear the class label and no need for more extraction/classification work. Isn’t it?

False class label means not related, isn’t it? As zero means no relation.

**_Andreas,_**

Your answers are correct. The multiple classification scheme for the chemprot task is as you explained.  

The false class label means that the candidate tags in the sentence are not related.


In sentence level, you have a limit max_seq_length=128, I think you can extend it but that depends on your setup. 

If you wish to apply the model on a huge dataset then I know that this also depends on your setup. For my experiments, I used google collaboratory and I had 150 abstracts.

I suggest you go with the first option: #1 divide each sentence alone and put all of them together in the testing.tsv file. The construction of a relation extraction dataset is a difficult task since you need expert annotators. 

You need to fine-tune the model on your dataset anyway to set the best hyperparameters #2 For fine-tuning, I need to provide a training data for just selecting the best hyperparameters.

Not sure why this happens. I do not think that they used validation dataset for the chemprot task #3 Why dev.tsv should be empty? So no need for the validation?


**### Me:**
I have still not understood why should I have to provide the label tags in the test data [index, sentence, label]?

The algorithm should predict that label, otherwise, I need to tag each sentence with it’s label which doesn’t make sense. 

I have tried to remove the label from the test data and it works fine, but would like to be sure?

In the training data, it is well understood that I have to provide the label tags for each sentence but for the test doesn’t make sense in my mind.

Could you please comment on this issue?

**### Andreas,**
The idea behind providing labels for the test.tsv is that the authors wanted to assess how well their model 'remembers' the labels. 

If you wish to just predict the test.tsv labels then you should not provide the label. You only need a very good training dataset.

If you see in the authors paper they provide results for RE task. The results they provide is a reflection on how well the model memorizes the labels they provide.

**Comment by LedaguenelArthur** [comment]
@JehadAldahdooh thx a lot !!

If I understand corectly the run_re.py script will (once trained obviously) predict if there's a relation between @GENE$ and @DISEASE$ for instance in a sentence :

"Allele A in @GENE$ gene +252 site can significantly increase the relative risk of @DISEASE$ in women in Guangdong, among which TNF-beta AA genotype might be one of the genetic susceptible factors for endometriosis."

I have a couple questions on this :

- What happens if the entry data has no @TAG$ ? Does the script produces an error ? Is the prediction always 0 ?
- Could we run such a script on a whole abstract or article or does it only work at the sentence level ?
- In your discussion Andreas and you mention that BioBERT performs "**relation classification task** and not **relation extraction** per se". Is there an easy way to tweak it into performing relation extraction ? An existing notebook or repo on that issue ? Do you suggest another tool to perform relation extraction ?
- What about relations with more than 2 entities ?

I know it's a lot of questions and you may not have all the answers, and I apologyze in advance if this is not the right place to ask for all of this (as you may have noticed I'm quite inexperienced with all of this and still learning how to maneuver these new tools).

Once again thx a lot for your answers,
Best regards,
Arthur Ledaguenel

**Comment by JehadAldahdooh** [comment]
Hi Arthur,

I may answer two of your questions quickly, others I can come to them later on 
Could we run such a script on a whole abstract or article or does it only work at the sentence level ?
The maximum sequence length was fixed to 512 tokens  which means you need to cut your text based with this max https://finetune.indico.io/chunk.html

What about relations with more than 2 entities ?
one possible way to do is to repeat this sentences with different entities
for example:
@gene$ and gene r increase the relative risk of @DISEASE$ 
gene t and @gene$ increase the relative risk of @DISEASE$ 

What happens if the entry data has no @TAG$ ? Does the script produces an error ? Is the prediction always 0 ?
I assume it is not relevant for the model since he doesn't know the entities so he will ignore it means no related


In your discussion Andreas and you mention that BioBERT performs "relation classification task and not relation extraction per se". Is there an easy way to tweak it into performing relation extraction ? An existing notebook or repo on that issue ? Do you suggest another tool to perform relation extraction ?

you can add one classification layer at the end of the model  binary or multi-class classification

hope you find it useful :)

**Comment by LedaguenelArthur** [comment]
Thx for being so quick and so thorough on your answer @JehadAldahdooh !!
